### DISCUSSION TITLE
route based async atom and auto refreshing

### DISCUSSION THREAD

**Discussion by lili21** [pmndrs/jotai#1147] [Ideas]
I would like the async atom refresh when navigate between route instead of just reading the outdated value. 
maybe we need a better support ( a utils function?) for this case.

**Comment by lili21** [comment]
can I reset a resolved async atom to initial state? in this way, the next `get` would trigger the refreshing

**Reply by dai-shi** [reply]
Does `atomWithDefault` with `useResetAtom` help?

**Reply by lili21** [reply]
https://codesandbox.io/s/youthful-burnell-82ewkc?file=/src/Home.jsx

it does't work as expected. looks like the useAtomValue be cached, and the get function never called again.

**Reply by dai-shi** [reply]
Oh, right. Reset only works when overwritten. That's my bad.

Now, I guess there are multiple workarounds for this.

One solution would be to use atomFamily and remove on unmount. But, this isn't very ideal as `remove` is not reactive. #1134 

Another solution is to use `refreshAtom`, but the result isn't very comfortable with React 18 strict effect behavior. https://codesandbox.io/s/mystifying-paper-izkixv

Maybe the canonical solution is to create an atom that is sync with router location. as in #1140 
Hm, that may still not be ideal. The question is whether if this can be declarative, or we need something imperative. Why do you want to refresh the atom in the first place?

**Reply by lili21** [reply]
let's say I got two page - a list page and a detail page. You can navigate between the two pages. and You can update some data at detail page, and after the updating, you navigate to list page. 
so the problem here is data/state sync， after you update an item's data at detail page, the changes should sync to the list. 

the simple solution is just refresh the list atom.
a better solution maybe is update the list atom manually.
and maybe a more better solution is combine those two  - update manually (good for UX) and refresh for data revalidating (good for consistent)

**Reply by dai-shi** [reply]
> You can update some data at detail page

Do you mean this is server data, not client data (like jotai atoms)?
Then, this could happen even if we are on a single page, right? (i.e. the issue is not router specific.)

**Reply by lili21** [reply]
yet, it's server data.
in general, the async atoms usually be used at router component, and router navigate is the time to clear the cache.

**Reply by lili21** [reply]
I kinda like the way remix handle the data  fetching - based on route.

**Reply by dai-shi** [reply]
jotai core has nothing to do with router conceptually. I have been thinking about a util function, an integration library or a 3rd-party library to work better with remix/nextjs/others, but no concrete ideas so far.

> router navigate is the time to clear the cache.

In this sense, jotai cache/store exists in a provider. So, you want to use Provider for each route.
https://codesandbox.io/s/nervous-resonance-ozw78z?file=/src/Home.jsx

**Reply by lili21** [reply]
the `home page data` didn't refresh, when navigate back from `about` page

**Reply by dai-shi** [reply]
Hmm, I thought it worked, but now confirmed not. Maybe it was my mistake.

**Reply by lili21** [reply]
maybe it's a data-fetching problem. and I should fix it at that level.

**Reply by dai-shi** [reply]
Hm, but atoms won't be reevaluated unless their deps change.

In general, I think this problem is important, and should be fixed or worked around.
I'm not sure how it should be done.
Do you want all atoms to be refreshed on route change, or just some of them?

**Reply by lili21** [reply]
just some of them.

**Reply by dai-shi** [reply]
Hm, my next question is how would you know/tell which some of them should be refreshed.

**Reply by lili21** [reply]
if we can somehow bind the async atoms to route, I guess we can clear it all based on route navigate. 
But IMO provide a utils function and let the users decide when and what to clear would be a good start.

**Reply by dai-shi** [reply]
Unlike what I thought in https://github.com/pmndrs/jotai/discussions/1147#discussioncomment-2712030,  I think this is router specific issue.
We basically have two options.
1. Declarative. Something like `atomWithLocation` #1140
2. Imperative. "clear" imperatively means creating new atoms in jotai, and `atomFamily.remove()` is one imperative methods. Of course, `useEffect` + reset is an option too (but not so happy as jotai way).

**Reply by dai-shi** [reply]
https://github.com/pmndrs/jotai/discussions/1147#discussioncomment-2758044

**Comment by dai-shi** [comment]
One problem is that there's no standard way to subscribe to route change. Related: #1140

**Comment by dai-shi** [comment]
https://codesandbox.io/s/objective-mccarthy-kx8jvu?file=/src/history.js

This is what I expect to work in a most jotai way. (Declarative)
Unfortunately, it's not working. `history` doesn't notify the route change... Not sure what's wrong.
